src/common/function/src/scalars/vector/distance.rs

```diff
@@ -0,0 +1,469 @@
+                                ),
+                            }
+                        );
+
+                        let f = <f32 as simsimd::SpatialSimilarity>::$similarity_method;
+                        // Safe: checked if the length of the vectors match
+                        let d = f(vec0.as_ref(), vec1.as_ref()).unwrap();
+                        result.push(Some(d));
+                    } else {
+                        result.push_null();
+                    }
+                }
+
+                return Ok(result.to_vector());
+            }
+        }
+
+        impl Display for $StructName {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(f, "{}", $display_name.to_ascii_uppercase())
+            }
+        }
+    }
+}
+
+define_distance_function!(CosDistanceFunction, "cos_distance", cos);
+define_distance_function!(L2SqDistanceFunction, "l2sq_distance", l2sq);
+define_distance_function!(DotProductFunction, "dot_product", dot);
+
+/// Parse a vector value if the value is a constant string.
+fn parse_if_constant_string(arg: &Arc<dyn Vector>) -> Result<Option<Vec<f32>>> {
+    if !arg.is_const() {
+        return Ok(None);
+    }
+    if arg.data_type() != ConcreteDataType::string_datatype() {
+        return Ok(None);
+    }
+    arg.get_ref(0)
+        .as_string()
+        .unwrap() // Safe: checked if it is a string
+        .map(parse_vector_from_string)
+        .transpose()
+}
+
+/// Convert a value to a vector value.
+/// Supported data types are binary and string.
+fn as_vector(arg: ValueRef<'_>) -> Result<Option<Cow<'_, [f32]>>> {
+    match arg.data_type() {
+        ConcreteDataType::Binary(_) => arg
+            .as_binary()
+            .unwrap() // Safe: checked if it is a binary
+            .map(|bytes| Ok(Cow::Borrowed(binary_as_vector(bytes)?)))
+            .transpose(),
+        ConcreteDataType::String(_) => arg
+            .as_string()
+            .unwrap() // Safe: checked if it is a string
+            .map(|s| Ok(Cow::Owned(parse_vector_from_string(s)?)))
+            .transpose(),
+        ConcreteDataType::Null(_) => Ok(None),
+        _ => InvalidFuncArgsSnafu {
+            err_msg: format!("Unsupported data type: {:?}", arg.data_type()),
+        }
+        .fail(),
+    }
+}
+
+/// Convert a u8 slice to a vector value.
+fn binary_as_vector(bytes: &[u8]) -> Result<&[f32]> {
+    if bytes.len() % 4 != 0 {
+        return InvalidFuncArgsSnafu {
+            err_msg: format!("Invalid binary length of vector: {}", bytes.len()),
+        }
+        .fail();
+    }
+
+    unsafe {
+        let num_floats = bytes.len() / 4;
+        let floats: &[f32] = std::slice::from_raw_parts(bytes.as_ptr() as *const f32, num_floats);
+        Ok(floats)
+    }
+}
+
+/// Parse a string to a vector value.
+/// Valid inputs are strings like "[1.0, 2.0, 3.0]".
+fn parse_vector_from_string(s: &str) -> Result<Vec<f32>> {
```

Comment:
I'm unsure whether we should directly support parsing vectors from strings, or if we should instead provide a UDF to handle the string-to-vector parsing.

Reply:
I'm comfortable with this concise usage unless some strong obstacle changes my mind.

Reply:
> I'm comfortable with this concise usage unless some strong obstacle changes my mind.

maybe rename to `parse_f32_vector_from_string` since it accepts valid f32 values only.